pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/StrictRealtimeSegmentAssignment.java

```diff
@@ -0,0 +1,108 @@
+      Map<InstancePartitionsType, InstancePartitions> instancePartitionsMap) {
+    Preconditions.checkState(instancePartitionsMap.size() == 1, "One instance partition type should be provided");
+    Map.Entry<InstancePartitionsType, InstancePartitions> typeToInstancePartitions =
+        instancePartitionsMap.entrySet().iterator().next();
+    InstancePartitionsType instancePartitionsType = typeToInstancePartitions.getKey();
+    InstancePartitions instancePartitions = typeToInstancePartitions.getValue();
+    Preconditions.checkState(instancePartitionsType == InstancePartitionsType.CONSUMING,
+        "Only CONSUMING instance partition type is allowed for table using upsert but got: " + instancePartitionsType);
+    _logger.info("Assigning segment: {} with instance partitions: {} for table: {}", segmentName, instancePartitions,
+        _tableNameWithType);
+    int segmentPartitionId =
+        SegmentAssignmentUtils.getRealtimeSegmentPartitionId(segmentName, _tableNameWithType, _helixManager,
+            _partitionColumn);
+    List<String> instancesAssigned = assignConsumingSegment(segmentPartitionId, instancePartitions);
+    List<String> idealAssignment = new ArrayList<>(instancesAssigned.size());
+    // Iterate the idealState to find the first segment that's in the same table partition with the new segment, and
+    // check if their assignments are same. We try to derive the partition id from segment name to avoid ZK reads.
+    Set<String> nonStandardSegments = new HashSet<>();
+    for (Map.Entry<String, Map<String, String>> entry : currentAssignment.entrySet()) {
+      LLCSegmentName llcSegmentName = LLCSegmentName.of(entry.getKey());
+      if (llcSegmentName == null) {
+        nonStandardSegments.add(entry.getKey());
+        continue;
+      }
+      if (llcSegmentName.getPartitionGroupId() == segmentPartitionId) {
+        idealAssignment.addAll(entry.getValue().keySet());
+        break;
+      }
+    }
+    if (idealAssignment.isEmpty()) {
+      _logger.debug("Check ZK metadata of segments: {} for any one from partition: {}", nonStandardSegments.size(),
+          segmentPartitionId);
+      // Check ZK metadata for segments with non-standard LLC segment names.
+      for (String nonStandardSegment : nonStandardSegments) {
+        if (SegmentAssignmentUtils.getRealtimeSegmentPartitionId(nonStandardSegment, _tableNameWithType, _helixManager,
+            _partitionColumn) == segmentPartitionId) {
+          idealAssignment.addAll(currentAssignment.get(nonStandardSegment).keySet());
+          break;
+        }
+      }
+    }
+    if (idealAssignment.isEmpty()) {
+      _logger.info("Found no existing assignment from idealState, using the one decided by instancePartitions");
+    } else if (!idealAssignment.equals(instancesAssigned)) {
+      _logger.info("Assignment: {} is inconsistent with idealState: {}, using the one as from idealState",
```

Comment:
This is at least WARN level, and we should consider emitting metrics because it means IS and instance partitions are async